### PR TITLE
Handle DB errors in categories search plugin

### PR DESCRIPTION
**Pull request by alikon**
updated to handle DB errors like #6612, #6632 and for #6591

Testing info
simulate a db error

### CODE REVIEW

File: plugins/search/categories/categories.php [hunk 1]
```diff
@@ -169,7 +169,15 @@ public function onContentSearch($text, $phrase = '', $ordering = '', $areas = nu
...
 		$db->setQuery($query, 0, $limit);
-		$rows = $db->loadObjectList();
+		try
+		{
+			$rows = $db->loadObjectList();
+		}
+		catch (RuntimeException $e)
+		{
+			$rows = array();
+			JFactory::getApplication()->enqueueMessage($e->getMessage(), 'error');
```
Comment (phproberto) [inline plugins/search/categories/categories.php]: What is the point in catching the error if we are going to then show it in a system message? I think it should be sent to logs or at least check that the user has core.admin rights before showing the DB error.

Comment (wilsonge) [inline plugins/search/categories/categories.php]: Well at least the page still renders :P But I agree with the point. Part of the issue is we don't want to show the database prefix to an average user. We need a more general error message and to ensure the DB failure is logged somewhere for admins

Comment (mbabker) [inline plugins/search/categories/categories.php]: JDatabaseDriver already has logging mechanisms in place...

Comment (phproberto) [inline plugins/search/categories/categories.php]: yes. I tested that today when @wilsonge suggested it in Glip. Is not easy to understand but you can get the debug system plugin showing that kind of errors.

log_databasequery

Is there a better way to do this?

Comment (phproberto) [inline plugins/search/categories/categories.php]: So for db exceptions we need to replace:

```php
JFactory::getApplication()->enqueueMessage($e->getMessage(), 'error');
```

with:

```php
if (JFactory::getUser()->authorise('core.admin'))
{
	JFactory::getApplication()->enqueueMessage($e->getMessage(), 'error');
}

Comment (wilsonge) [inline plugins/search/categories/categories.php]: Would it not be better to do that at the JModuleHelper level?

Comment (phproberto) [inline plugins/search/categories/categories.php]: If it's at the module rendering level I think it's not enough:

* Is not something we can use everywhere (a recommended practice that we can document)
* Sometimes you want to return content even having an exception that requires admin review

Comment (mbabker) [inline plugins/search/categories/categories.php]: It isn't JModuleHelper's job to act as a global exception handler for
modules.  Same as JComponentHelper doesn't for components.

The error message displayed to the user, if one is displayed, should
always be sanitized.  IMO there isn't a need to wrap the enqueueMessage
call in a core.admin permission check.  Use the logs for verbose error
logging with "sensitive" data..  In the UI, it is difficult to turn one on,
I agree, but the code structure is already present to use this stuff.
Let's not make things more difficult than we need to please.

Comment (Bakual) [inline plugins/search/categories/categories.php]: I agree with Michael. Don't show the query error to the user. Just show a generic message "Oops, the database query failed. Please notifiy the administrator." Or something like that. I wouldn't add ACL checks for that kind of things.
We could add docs for the administrator how to get the full error.

**Comment by sovainfo** [comment]
The errors should be logged and currently that requires proper configuration of debug. Consider that not sufficient. The product should have a logger enabled for errors.

**Comment by mbabker** [comment]
It can be default enabled but there are conditions to keep in mind:

- A hosting platform where Joomla can't write log files (due to poor config
or a user implementing extra "security" measures)
- Failing to write to the log shouldn't trigger a fatal error (I think I've
seen instances where this may happen, in part due to the lack of
permissions)
- Reviewing where we are actually using JLog now.  We have a log callback
that enqueues messages for the 'jerror' log category (which IMO is really
not optimal; if we want to log and display a message just call both...).
We don't want to be writing logs on every request a site gets; those files
blow up pretty quickly, especially if catching deprecations or logging all
database queries.

**Comment by sovainfo** [comment]
Guess you missed the second word: ERROR
Added this to cms.php:
JLog::addLogger(array(), JLog::ERROR, array('deprecated'), true);
Then created an error in the featured query, it was logged and embarrassingly enough also shown:
component error handling

**Comment by alikon** [comment]
the point is how the core modules,plugins,ect handle db errors  ( #6591 )  ?

showing a generic message 
```php
JFactory::getApplication()->enqueueMessage('blah blah ', 'warning');
```
or  something different like 
```php
if (JFactory::getUser()->authorise('core.admin'))
{
 JLog::addLogger("blla bla");  
........
}
```
i would prefer the keep it simplest stupid "bla bla" msg

but not sure to understand what exactly is the "family feeling" 

i'm missing something?

p.s
a special thanks to all the tester

**Comment by wilsonge** [comment]
Me and @roland-d just had a chat about this. We have made an executive call that we show a generic "blah blah" message to users and then just add the exception message into the Logs. Just make sure it isn't in a category that will be shown to the frontend user.

**Comment by alikon** [comment]
done
i've used a ready lang string  maybe something different?

**Comment by wilsonge** [comment]
I think probably something different. I'll ping @brianteeman in the language string group on glip

**Comment by roland-d** [comment]
@wilsonge Any update on this?

**Comment by wilsonge** [comment]
@brianteeman recommended as it's a string to frontend users we either display a hugely simplified string of "There has been an error" or don't show a string at all.

File: plugins/search/categories/categories.php [hunk 2]
```diff
@@ -169,7 +169,16 @@ public function onContentSearch($text, $phrase = '', $ordering = '', $areas = nu
 		}
 
 		$db->setQuery($query, 0, $limit);
-		$rows = $db->loadObjectList();
+		try
```
Comment (roland-d) [inline plugins/search/categories/categories.php]: @alikon Can you please add an empty line before try so we stick to the codestyle. Thanks.

**Comment by roland-d** [comment]
@test Works as described, after generating an error, I get a simple error message that the database connection is not there.

To test, I changed the plugins/search/categories/categories.php file on line 152 and changed THEN to THN. On the frontend, I used the search function to search categories only and received the error page with a large DB query. After applying the patch and making the same change, I now receive a graceful message.

**Comment by alikon** [comment]
@roland-d  i'm sorry for my pr patchworks
look at this one please #7073

**Comment by roland-d** [comment]
@alikon There are a lot of these db PRs. Is there a comprehensive list with all of them that should be tested?

**Comment by alikon** [comment]
@roland-d  my search string 
```
is:open is:pr author:alikon Handle DB 
```
;)

**Comment by designbengel** [comment]
Did same testing procedure like @roland-d - successfull<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6710">issues.joomla.org/joomla-cms/6710</a>.</sub>

**Comment by zero-24** [comment]
RTC :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6710">issues.joomla.org/joomla-cms/6710</a>.</sub>

**Comment by wilsonge** [comment]
Closing in favour of https://github.com/joomla/joomla-cms/pull/7073

**Pull request closed without merging**